design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,84 @@
+KubeVirt components.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+### Pitfalls of the current solution
+
+We are facing multiple challenges with the current solution:
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
+
+- Complex to collect all existing metrics and alerts for automatic documentation
+
+- High levels of intertwined monitoring and business logic code
+
+### Goals
+
+- Have a centralized place for the creation of monitoring resources
+- Have a consistent monitoring package between all projects
+- Have a code struct that is easy to maintain and evolve
+- Move closer to the Kubernetes metric implementation style
+- Have a structure and tools to accurately generate monitoring documentation,
+lint metrics and alerts, and define allow-list, deny-list and opt-in metrics
+
+### Non-Goals
+
+- Update any metrics or alerts
+- Apply these ideas to events in this phase
+- Create a new component to handle all project monitoring
+
+## How
+
+- Create a new separate “monitoring” package, in the "internal" folder,
+in each project with all the metrics and alerts, and the logic to both register
+them and set their values
+
+- Use custom metrics wrapper for metrics to allow custom fields like stability
+and deprecation status
+
+- Update the monitoring documentation generator
+
+- Add unit tests to check metrics are being added to the Prometheus registry and
+PrometheusRule contains all alerts
+
+![code struct](struct.png)
+
+### Example implementation in [operator-sdk](https://github.com/operator-framework/operator-sdk) example operator
+
+Monitoring commit:
+https://github.com/machadovilaca/operator-sdk/commit/8e921f951d0445b83d9a0937edb39c365a384bb4
+
+Full example:
+https://github.com/machadovilaca/operator-sdk/tree/8e921f951d0445b83d9a0937edb39c365a384bb4/testdata/go/v4-alpha/monitoring/memcached-operator-external-package
+
+## Action Plan
+
+- Refactor the implementation of the metrics in KubeVirt
+- Implement the metrics changes in other components
+- Refactor the implementation of the recording rules and alerts in KubeVirt
+- Implement the recording rules and alerts changes in other components
+- Add linter to the library package
+- Implement allow-list, deny-list and opt-in metrics logic
+
+## Open Questions
+
+- Should we add a mechanism to prevent developers to create and register metrics
```

Comment:
This depends on the build system.
But if the structure is clear, contributors will likely add it to the place it makes sense. Most will just follow the current norm, unless they think it is totally wrong.
I would worry about this only if there is proof that it is a problem, and not before.